keyboards/sevencrumbs/matrix/keymaps/via/keymap.c

```diff
@@ -0,0 +1,127 @@
+// Copyright 2023 Chris H <chris@zoid.com.au>
+// SPDX-License-Identifier: GPL-3.0-only
+
+#include QMK_KEYBOARD_H
+#include "eeprom.h"
+#include "color.h"
+#include "backlight.h"
+
+typedef struct {
+	uint8_t hue;
+	uint8_t saturation;
+} color_t;
+
+typedef struct {
+	uint8_t brightness;
+	color_t color[RGB_MATRIX_LED_COUNT];
+} config_t;
+
+static_assert(VIA_EEPROM_CUSTOM_CONFIG_SIZE == sizeof(config_t), "VIA EEPROM custom configuration size must match configuration size");
+
+enum values {
+	brightness = 1,
+	color = 2
+};
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+	LAYOUT(
+		KC_1, KC_2,
+		KC_3, KC_4
+	)
+};
+
+static bool initialized;
+static bool suspended;
+static config_t config = {.brightness = 255};
+
+void config_save(void) {
+	eeprom_update_block(&config, (void*) VIA_EEPROM_CUSTOM_CONFIG_ADDR, VIA_EEPROM_CUSTOM_CONFIG_SIZE);
+}
+
+void eeconfig_init_user(void) {
+	config_save();
+}
+
+void backlight_update(void) {
+	if (!initialized) {
+		backlight_init();
+		eeprom_read_block(&config, (void*) VIA_EEPROM_CUSTOM_CONFIG_ADDR, VIA_EEPROM_CUSTOM_CONFIG_SIZE);
+		initialized = true;
+	}
+
+	if (suspended) {
+		backlight_set_color_all(0, 0, 0);
+	} else {
+		HSV color_hsv = {.v = config.brightness};
+
+		for (int index = 0; index < RGB_MATRIX_LED_COUNT; index++) {
+			color_t color_hs = config.color[index];
+
+			color_hsv.h = color_hs.hue;
+			color_hsv.s = color_hs.saturation;
+
+			RGB color_rgb = hsv_to_rgb(color_hsv);
+
+			backlight_set_color(index, color_rgb.r, color_rgb.g, color_rgb.b);
+		}
+	}
+
+	backlight_flush();
+}
+
+void suspend_power_down_user(void) {
+	suspended = true;
+	backlight_update();
+}
+
+void suspend_wakeup_init_user(void) {
+	suspended = false;
+	backlight_update();
+}
+
+void via_custom_value_command_kb(uint8_t *data, uint8_t length) {
+	uint8_t command = data[0];
+
+	if (data[1] == id_custom_channel) {
+		if (command == id_custom_save) {
+			config_save();
+			return;
+		}
+
+		if (data[2] == brightness) {
+			if (command == id_custom_set_value) {
+				config.brightness = data[3];
+				backlight_update();
+				return;
+			}
+
+			if (command == id_custom_get_value) {
+				data[3] = config.brightness;
+				return;
+			}
+		}
+
+		if (data[2] == color) {
+			uint8_t index = data[3];
+
+			if (index < RGB_MATRIX_LED_COUNT) {
+				color_t *color = config.color + index;
+
+				if (command == id_custom_set_value) {
+					color->hue = data[4];
+					color->saturation = data[5];
+					backlight_update();
+					return;
+				}
+
+				if (command == id_custom_get_value) {
+					data[4] = color->hue;
+					data[5] = color->saturation;
+					return;
+				}
+			}
+		}
+	}
+
+	data[0] = id_unhandled;
+}
```

Comment:
This custom VIA code is currently not being accepted due to posing a maintenance burden and being scattered throughout the repo. 
```suggestion
```

Reply:
How would you suggest I implement this functionality? I'm looking to get the keyboard definition into VIA - which requires acceptance into QMK. As the macro pad is symmetrical, per-key backlight customisation is essential to differentiate the buttons.

Reply:
It's a problem that should and needs to be solved in the core, basically.  

Otherwise, we *will* have a dozen different implementations for a dozen different boards, something that we already started to see, hence the "no".  Especially as it's very possible (likely) that core changes could break custom implementations. 

Unfortunately, it means that somebody needs to do the work to get it in core.

Reply:
I'd love to see per-key backlight support in the core, but I'm not aware of any features in VIA that would work for more than a handful of keys. I'm relatively lucky that I only have four buttons, which means VIA's custom UI works really well in my case. Are there any other keyboards with customisable backlight colours? The closest I could find was a couple with indicator colour customisation.

I'm happy to take the risk of core changes breaking my implementation - I'll have to maintain the functionality regardless of whether it gets accepted, as it's an essential feature for the product. Ideally, I'd like the firmware available in the QMK repository to be identical to the one shipped with the device.